tests/thumb/thumb.armhf.py

```diff
@@ -0,0 +1,109 @@
+import logging
+
+import smallworld
+from smallworld.state.cpus.arm import ARM
+
+# setup logging and hinting
+smallworld.logging.setup_logging(level=logging.INFO)
+
+THUMB_BLOCK_OFFSET = 0x10
+
+for arch in [
+    smallworld.platforms.Architecture.ARM_V7A,
+    smallworld.platforms.Architecture.ARM_V7M,
+    smallworld.platforms.Architecture.ARM_V7R,
+]:
+    # harness definition
+    platform = smallworld.platforms.Platform(
+        arch, smallworld.platforms.Byteorder.LITTLE
+    )
+    emulator = smallworld.emulators.UnicornEmulator(platform)
+    machine = smallworld.state.Machine()
+    cpu: ARM = smallworld.state.cpus.CPU.for_platform(platform)
+    code = smallworld.state.memory.code.Executable.from_filepath(
+        "thumb/thumb.armel.bin", address=0x1000
+    )
+    machine.add(cpu)
+    machine.add(code)
+    machine.add_exit_point(code.address + code.get_capacity())
+
+    # test step_instruction starting with ARM
+    cpu.pc.set(code.address)
+    for step in machine.step(emulator):
+        pass
+    cpu.r0.extract(emulator)
+    print(f"STEP_{arch.name}={hex(cpu.r0.get())}")
+
+    # test step_instruction starting with Thumb
+    cpu.pc.set(code.address + THUMB_BLOCK_OFFSET)
+    emulator.set_thumb()
+    cpu.r0.set(0)
+    for step in machine.step(emulator):
+        pass
+    cpu.r0.extract(emulator)
+    print(f"STEP_{arch.name}={hex(cpu.r0.get())}")
+
+    # test step_block starting with ARM
+    cpu.pc.set(code.address)
+    cpu.r0.set(0)
+    machine.apply(emulator)
+    emulator.step_block()
+    emulator.step_block()
+    emulator.step_block()
+    cpu.r0.extract(emulator)
+    print(f"BLOCK_{arch.name}={hex(cpu.r0.get())}")
+
+    # test step_block starting with Thumb
+    cpu.pc.set(code.address + THUMB_BLOCK_OFFSET)
+    emulator.set_thumb()
+    cpu.r0.set(0)
+    machine.apply(emulator)
+    emulator.step_block()
+    emulator.step_block()
+    cpu.r0.extract(emulator)
+    print(f"BLOCK_{arch.name}={hex(cpu.r0.get())}")
+
+    # test run starting with ARM
+    cpu.pc.set(code.address)
+    cpu.r0.set(0)
+    machine.emulate(emulator)
+    cpu.r0.extract(emulator)
+    print(f"RUN_{arch.name}={hex(cpu.r0.get())}")
+
+    # test run starting with Thumb
+    cpu.pc.set(code.address + THUMB_BLOCK_OFFSET)
+    emulator.set_thumb()
+    cpu.r0.set(0)
+    machine.emulate(emulator)
+    cpu.r0.extract(emulator)
+    print(f"RUN_{arch.name}={hex(cpu.r0.get())}")
+
+    # test ISA persistance across emulator instances
+    emulator2 = smallworld.emulators.UnicornEmulator(platform)
+    cpu.pc.set(code.address + THUMB_BLOCK_OFFSET)
+    emulator.set_thumb()
+    cpu.r0.set(0)
+    machine.step(emulator)
```

Comment:
Generator does nothing if you don't `next()` it 😢 
It would be nice if there was a way to have a warning for this. I've written code with futures before where they emit a warning if they get GC'd without ever having been waited upon, maybe we could do something like that here?